docs: Service Worker examples and explanations (#15595)
microsoft/playwright: docs/src/service-workers-guide.md

---
id: service-workers-guide
title: "Service Workers Guide"
---

## Disabling Service Workers

[Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) provide a browser-native way to handle requests made with the native [Fetch API (`fetch`)](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

> If you're using Playwright's [Network-related features](./network.md) like the routing APIs ([`method: BrowserContext.route`], [`method: BrowserContext.routeFromHAR`], [`method: Page.route`], [`method: Page.routeFromHAR`]), or rely heavily on Network Events (e.g. [`event: Page.request`], [`event: BrowserContext.request`], etc.), we recommend disabling Service Workers by setting [`option: Browser.newContext.serviceWorkers`] to `'block'`.

Author: I'm not sure I agree with this recommendation. The qualification helps, but there is a point to be made this potentially changes the behavior of the app under test s.t. it no longer reflects what a user would experience.

Reviewer: I think the paragraph about FF/WK is the most important one. If you use networking APIs and test in these browsers, we strongly recommend to disable SW.

  page.goto('example-with-a-service-worker.html'),
]);
```

The [`BrowserContext.serviceworker`] event is emitted ***before*** the Service Worker main script has been evaluated, so before calling [ServiceWorker.evaluate] you should wait on its activation:

```js
await expect.poll(() => serviceWorker.evaluate(() => self.registration.active?.state)).toBe('activated');
```

## Network Events and Routing

A Network Request made within a Service Worker (along with the main script request itself) will have:

* [`event: BrowserContext.request`] event emitted
* [`method: BrowserContext.route`] will see the request
* [`method: Request.serviceWorker`] will be set to the [Service Worker] instance, and [`method: Request.frame`] will **throw**
* [`method: Response.fromServiceWorker`] will return `false`

A Network Request made within a page will have:

* [`event: BrowserContext.request`] event emitted
* [`event: Page.request`] event emitted
* [`method: BrowserContext.route`] and [`method: Page.route`] will **not** see the request (if a Service Worker's fetch handler was registered)
* [`method: Request.serviceWorker`] will be set to `null`, and [`method: Request.frame`] will return the [Frame]

> * [`method: Response.fromServiceWorker`] will return `true` (if a Service Worker's fetch handler was registered)

Author: Hopefully this (coupled with the tables below) make it clearer and more closely align with the link you shared.

Given enough context, the test case I added also seems elucidating, but I don't think it belongs in the docs.

---
id: service-workers-guide
title: "Service Workers Guide"
---

> 

Reviewer: Suggesting the following intro (rough sketch):

[Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) provide a browser-native way to handle requests made with the native [Fetch API (`fetch`)](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

Depending on your testing needs, here are the scenarios supported by Playwright:
1. Test your page with a service worker. This works out of the box in all supported browsers.
2. Test your page without a service worker. Use `serviceWorkers: 'block'` option.
3. Route network traffic, whether it comes from a service worker or not. In Firefox and WebKit, use `serviceWorkers: 'block'` option to avoid service worker network traffic entirely. In Chromium, either block service workers or use [`BrowserContext.route()`].
4. Test your service worker implementation. Only in Chromium, use [`BrowserContext.serviceWorkers()`] to get access to the service worker and evaluate there.

## Disabling Service Workers

[Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) provide a browser-native way to handle requests made with the native [Fetch API (`fetch`)](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API).

If you're using Playwright's [Network-related features](./network.md) like the routing APIs ([`method: BrowserContext.route`], [`method: BrowserContext.routeFromHAR`], [`method: Page.route`], [`method: Page.routeFromHAR`]), or rely heavily on Network Events (e.g. [`event: Page.request`], [`event: BrowserContext.request`], etc.), we recommend disabling Service Workers by setting [`option: Browser.newContext.serviceWorkers`] to `'block'`.

Unless you're testing the Service Worker behavior itself, blocking ensures all requests are routed to your code via Playwright. It also simplifies Network-related APIs within Playwright.

> If you need to visibility or routing into all Network events in Firfox or WebKit, currently `'block'`ing Service Workers is required.

Author: ```suggestion
If you need visibility or routing into all Network events in Firfox or WebKit, currently `'block'`ing Service Workers is required.
```

## Service Worker Fetch

:::note
The next sections are only currently supported when using Playwright with Chrome/Chromium.
:::

### Accessing Service Workers and Waiting for Activation

> You can use [`method: BrowserContext.serviceWorkers`] to list the Service Workers, or specifically watch for it:

Author: ```suggestion
You can use [`method: BrowserContext.serviceWorkers`] to list the Service Workers, or specifically watch for the creation of a Service Worker if you expect one to be created:
```

```js
const [ serviceWorker ] = await Promise.all([
  context.waitForEvent('serviceworker'),
  page.goto('example-with-a-service-worker.html'),
]);
```

> The [`BrowserContext.serviceworker`] event is emitted ***before*** the Service Worker main script has been evaluated, so before calling [ServiceWorker.evaluate] you should wait on its activation:

Author: ```suggestion
The [`event: BrowserContext.serviceWorker`] event is emitted ***before*** the Service Worker main script has been evaluated, so before calling [`method: ServiceWorker.evaluate`] you should wait on its activation:
```

```js

> await expect.poll(() => serviceWorker.evaluate(() => self.registration.active?.state)).toBe('activated');

Author: We probably ought to add an event/PW API for this: something like `serviceWorker.waitForLifecycle('active')`, and/or `serviceWorker.waitForActive()`.

Reviewer 2: can this be achieved using something like this?
```ts
expect.poll(() => page.evaulaute(() => navigator.serviceWorker.controller));
```

Author: A variant of that could be used. I was hoping to keep it on the SW PW object instead of the Page, but looks like we don't have generic poll in other language ports, so I've updated the workaround for now.

```

## Network Events and Routing

> A Network Request made within a Service Worker (along with the main script request itself) will have:

Reviewer: ```suggestion
Any network request made by the Service Worker will have:
```

* [`event: BrowserContext.request`] event emitted
* [`method: BrowserContext.route`] will see the request
* [`method: Request.serviceWorker`] will be set to the [Service Worker] instance, and [`method: Request.frame`] will **throw**
* [`method: Response.fromServiceWorker`] will return `false`

> A Network Request made within a page will have:

Reviewer: I'd start this section with something like this:

In a typical service worker that has a fetch handler and responds with a fetch (an example source you had would be handy), each page request generates an additional request issued by the Service Worker.

* [`event: BrowserContext.request`] event emitted
* [`event: Page.request`] event emitted
* [`method: BrowserContext.route`] and [`method: Page.route`] will **not** see the request (if a Service Worker's fetch handler was registered)
* [`method: Request.serviceWorker`] will be set to `null`, and [`method: Request.frame`] will return the [Frame]
* [`method: Response.fromServiceWorker`] will return `true` (if a Service Worker's fetch handler was registered)

Consider the following Service Worker Main Script:

```js

> // filename: service-worker-main.js

Reviewer 2: this is duplicated in `service-worker-main.js`
can you use here `<!-- filename: service-worker-main.js -->` ?

Author: It's a JS example, so want to keep it as JS syntax. The test code is actually independent from this.

      if (response) return response;

      // 2. Re-write request for /foo to /bar
      if (event.request.url.endsWith("foo")) return fetch("./bar");

      // 3. Prevent `tracker.js` from being retrieved, and returns a placeholder response
      if (event.request.url.endsWith("tracker.js"))
        return new Response('conosole.log("no trackers!")', {
          status: 200,
          headers: { "Content-Type": "text/javascript" },
        });

      // 4. Otherwise, fallthrough, perform the fetch and respond
      return fetch(event.request);
    })()
  );
});

self.addEventListener("activate", (event) => {
  event.waitUntil(clients.claim());
});
```

And a Page that simply registers the Service Worker:

```html
<!-- filename: index.html -->
<script>
  window.registrationPromise = navigator.serviceWorker.register('./service-worker-main.js');
</script>
```

On the first visit to the page via [`method: Page.goto`], the following Request/Response events would be emitted:

| Scope   | Target         | URL                       | Routed | fromServiceWorker |
| -       | -              | -                         | -      | -                 |

> | Context | Frame          | index.html                | Yes    |                   |

Reviewer: "Scope" is a bit unclear, can we do it like `page.on('request')` vs `context.on('request')`?

Author: Agreed, just updated. See screenshot for what it looks like. I also changed the `Scope` heading to `Event` since we then put events underneath it.

await page.evaluate(() => fetch('./foo'));
await page.evaluate(() => fetch('./tracker.js'));
await page.evaluate(() => fetch('./fallthrough.txt'));
```

The following Request/Response events would be emitted:

| Scope   | Target         | URL                       | Routed | fromServiceWorker |
| -       | -              | -                         | -      | -                 |
| Context | Frame          | addressbook.json          |        | Yes               |
| Page    | Frame          | addressbook.json          |        | Yes               |
| Context | Service Worker | bar                       | Yes    |                   |
| Context | Frame          | foo                       |        | Yes               |
| Page    | Frame          | foo                       |        | Yes               |
| Context | Frame          | tracker.js                |        | Yes               |
| Page    | Frame          | tracker.js                |        | Yes               |
| Context | Service Worker | fallthrough.txt           | Yes    |                   |
| Context | Frame          | fallthrough.txt           |        | Yes               |

> | Page    | Frame          | fallthrough.txt           |        | Yes               |

Author: I think this table makes it much clearer (and it's generate from a Test nearly verbatim):

<img width="760" alt="Screen Shot 2022-07-13 at 2 12 17 AM" src="https://user-images.githubusercontent.com/11915034/178699293-8c4d08d0-ad43-4253-a146-90325deda97d.png">
